core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -117,9 +174,59 @@ class TransactionCoordinator(brokerId: Int,
     txnManager.removeTransactionsForPartition(transactionStateTopicPartitionId)
   }
 
+  def tryUpdateTxnState(transactionalId: String,
+                        newMetadata: TransactionMetadata,
+                        responseCallback: TxnMetadataUpdateCallback): Unit = {
+
+    def completeUpdateCallback(error: Errors): Unit = {
+      // first respond to the clients, then try to complete one other delayed metadata update
+      // operations waiting on this transactional id
+      responseCallback(error)
+
+      txnUpdatePurgatory.checkAndCompleteOne(transactionalId)
+    }
+
+    // try to update the transaction metadata and append the updated metadata to txn log;
+    // if there is no such metadata treat it as invalid pid mapping error.
+    txnManager.getTransaction(transactionalId) match {
+      case None =>
+        completeUpdateCallback(Errors.INVALID_PID_MAPPING)
+
+      case Some(metadata) =>
+        val error = metadata synchronized {
+          if (metadata.pid != newMetadata.pid) {
+            Errors.INVALID_PID_MAPPING
+          } else if (metadata.epoch != newMetadata.epoch) {
+            Errors.PRODUCER_FENCED
+          } else if (metadata.hasPendingTransition) {
+            // using a place-holder error code to indicate parking the request to purgatory,
+            // instead of responding to client with the error code
+            Errors.COORDINATOR_LOAD_IN_PROGRESS
+          } else if (!metadata.prepareTransitionTo(newMetadata.state)) {
+            Errors.INVALID_TXN_STATE
+          } else {
```

Comment:
Oops, I had that in the previous commit but the rebase got it away.. adding it back!